external-crates/move/crates/move-compiler/src/typing/deprecations.rs

```diff
@@ -0,0 +1,765 @@
+                "ICE: Deprecations are not allowed in '{}'s and should have already been checked",
+                location
+            ),
+        };
+
+        for (DeprecationId { deprecation, .. }, locs) in self.deprecation_warnings {
+            let mut locs: Vec<_> = locs
+                .into_iter()
+                .flat_map(|(position, locs)| {
+                    Self::merge_locations(locs.into_iter().collect())
+                        .into_iter()
+                        .map(move |loc| (position, loc))
+                })
+                .collect();
+            locs.sort_by_key(|(_, loc)| *loc);
+            locs.reverse();
+
+            let initial_diag = locs.pop().expect("ICE: locs should not be empty");
+
+            let attr_deprecation = attr_deprecation(deprecation.location);
+            let location_string = match deprecation.location {
+                AttributePosition::Module => format!(
+                        "This {} is deprecated since the whole module that it is declared in is marked deprecated",
+                        initial_diag.0
+                    ),
+                x => format!("This {} is deprecated", x),
+            };
+
+            let initial_message = match deprecation.deprecation_info {
+                DeprecationInfo::Deprecated => location_string,
+                DeprecationInfo::DeprecatedWithNote(note) => format!("{location_string}: {note}"),
+            };
+
+            let mut diag = diag!(attr_deprecation, (initial_diag.1, initial_message));
+
+            for (position, loc) in locs {
+                diag.add_secondary_label((loc, format!("Deprecated {position} used here")));
+            }
+            self.env.add_diag(diag);
+        }
+    }
+
+    // Merge contiguous locations into single locations within `locs`.
+    fn merge_locations(mut locs: Vec<Loc>) -> Vec<Loc> {
+        locs.sort();
+        locs.reverse();
+
+        let mut merged = vec![];
+        let mut current = locs.pop().expect("ICE: locs should not be empty");
+        let mut start = current.start();
+        let mut end = current.end();
+
+        while let Some(next) = locs.pop() {
+            if current.file_hash() != next.file_hash() || end < next.start() {
+                merged.push(Loc::new(current.file_hash(), start, end));
+                current = next;
+                start = current.start();
+                end = current.end();
+            } else {
+                end = std::cmp::max(end, next.end());
+            }
+        }
+        merged.push(Loc::new(current.file_hash(), start, end));
+
+        merged
+    }
+}
+
+// Entrypoint: Processes the `prog` and adds deprecation warnings to the `context`.
+pub fn add_deprecation_warnings(env: &mut CompilationEnv, prog: &mut T::Program) {
+    let mut deprecations = Deprecations::new(env, prog);
+    deprecations.visit(prog);
+    deprecations.process_and_add_deprecation_warnings();
```

Comment:
I would group the deprecation warnings by each module member. In other words, we would make a new context when visiting a module member, and process any warnings when we are done visiting. I think that will make for the eaisest to digest warnings, since I think that is how folks tend to conceptualize and break down their code. 
The effect is if I call `deprecated_f` in `bar` and `foo`, I would get a separate warning in `bar` and in `foo`. 

Alternative to making a separate context for each member would be to make a big ol map
```
collected_warnings: Map</* usage site */ ModuleMember<Option<Name>>, Map</* deprecated item */, ModuleMember<Option<Name>>, Vec<Loc>>>
```

Reply:
The latter is exactly what we do today:) With the `DeprecationId` we are grouping the same warning in the same context together, but then split different deprecation warnings in the same context out into different top-level warnings. E.g., if I had

```
#[deprecated]
fun bar() { }

#[deprecated]
fun foo() { }

fun caller() {
   foo();
   bar();
   foo(); 
}

fun other_caller() {
   foo();
}
```

This would generate 3 top-level warnings:
* one for the two calls to `foo` inside of `caller`. 
* one for the call to `bar` inside of `caller`.
* one for the call to `foo` inside of `other_caller`.